### ISSUE TITLE
Incremental backup - enabling delta checksum because of fsm files out of sync on replica

### ISSUE THREAD

**Issue by krmozejko** [pgbackrest/pgbackrest#2020]
Please provide the following information when submitting an issue (feature requests or general comments can skip this):

1. pgBackRest version: 2.44

2. PostgreSQL version: 11.14

3. Operating system/version - if you have more than one server (for example, a database server, a repository host server, one or more standbys), please specify each:
- repository: 20.04.5 LTS
- primary db server: 18.04.6 LTS
- replica db server: 18.04.6 LTS

4. Did you install pgBackRest from source or from a package? 
from package

5. Please attach the following as applicable:
    - `pgbackrest.conf` file(s)
    - `postgresql.conf` settings applicable to pgBackRest (`archive_command`, `archive_mode`, `listen_addresses`, `max_wal_senders`, `wal_level`, `port`)
    - errors in the postgresql log file before or during the time you experienced the issue
    - log file in `/var/log/pgbackrest` for the commands run (e.g. `/var/log/pgbackrest/mystanza_backup.log`)

```
# pgbackrest-repository.conf
[prod]
pg1-host=10.114.0.25
pg1-path=/var/lib/postgresql/11/main
pg2-host=10.114.0.24
pg2-path=/var/lib/postgresql/11/main


[global]
backup-standby=y
process-max=16
repo1-path=/var/lib/pgbackrest
repo1-retention-full=1
start-fast=y
log-level-console=detail
log-level-file=debug
```
```
# pgbackrest-dbserver.conf
[prod]
pg1-path=/var/lib/postgresql/11/main

[global]
log-level-console=detail
log-level-file=debug
repo1-host=10.114.0.32
```
```
# postgresql.conf
wal_level = 'logical'
archive_mode = on
archive_command = 'pgbackrest --stanza=prod archive-push %p'
max_wal_senders = 10
```
```
# incremental backup log

pgbackrest --stanza=prod --type=incr backup
[Step 1/1] 2023-03-06 21:00:06.474 P00   INFO: backup command begin 2.44: --backup-standby --exec-id=2911970-9abb81ca --log-level-console=detail --log-level-file=debug --pg1-host=10.114.0.24 --pg2-host=10.114.0.25 --pg1-path=/var/lib/postgresql/11/main --pg2-path=/var/lib/postgresql/11/main --process-max=16 --repo1-path=/var/lib/pgbackrest --repo1-retention-full=1 --stanza=prod --start-fast --type=incr
[Step 1/1] 2023-03-06 21:00:08.263 P00   INFO: last backup label = 20230306-094058F_20230306-150011I, version = 2.44
[Step 1/1] 2023-03-06 21:00:08.263 P00   INFO: execute non-exclusive backup start: backup begins after the requested immediate checkpoint completes
[Step 1/1] 2023-03-06 21:00:10.669 P00   INFO: backup start archive = 00000005000018E8000000BF, lsn = 18E8/BF0039D0
[Step 1/1] 2023-03-06 21:00:10.669 P00   INFO: wait for replay on the standby to reach 18E8/BF0039D0
[Step 1/1] 2023-03-06 21:00:12.874 P00   INFO: replay on the standby reached 18E8/BF0039D0
[Step 1/1] 2023-03-06 21:00:12.874 P00   INFO: check archive for prior segment 00000005000018E8000000BE
[Step 1/1] WARN: file 'base/3084743/3085598_fsm' has same timestamp (1677435258) as prior but different size (prior 32768, current 65536), enabling delta checksum
```

6. Describe the issue:
When I run incremental backup with `--backup-standby` option enabled I get above warning. In result of enabling delta checksum backup takes much longer - so long that it undermines sense of using incremental instead of full. It seems that this warning is caused by difference between primary and replica:
```
# primary
~$ stat /var/lib/postgresql/11/main/base/3084743/3085598_fsm 
  File: /var/lib/postgresql/11/main/base/3084743/3085598_fsm
  Size: 65536           Blocks: 128        IO Block: 4096   regular file
Device: 811h/2065d      Inode: 4587946     Links: 1
Access: (0600/-rw-------)  Uid: (  107/postgres)   Gid: (  114/postgres)
Access: 2023-03-07 09:12:06.226219123 +0100
Modify: 2023-02-26 19:14:18.544143085 +0100
Change: 2023-02-26 19:14:18.544143085 +0100
 Birth: -
```
```
# replica
~$ stat /var/lib/postgresql/11/main/base/3084743/3085598_fsm
  File: /var/lib/postgresql/11/main/base/3084743/3085598_fsm
  Size: 32768           Blocks: 64         IO Block: 4096   regular file
Device: 811h/2065d      Inode: 3670061     Links: 1
Access: (0600/-rw-------)  Uid: (  107/postgres)   Gid: (  114/postgres)
Access: 2023-03-06 09:56:32.521505695 +0100
Modify: 2022-08-26 11:11:19.339582876 +0200
Change: 2022-08-26 11:11:19.339582876 +0200
 Birth: -
```

I checked other databases and there is more *_fsm files which are smaller or even do not exist on replica. Above modify date `2022-08-26` is when the replica was created. I used pg_basebackup and then started postgres with `recovery.conf` to set it up. 
```
# recovery.conf
standby_mode = 'on'
primary_conninfo = 'host=10.114.0.25 user=repmgr application_name=master port=5432'
recovery_target_timeline = 'latest'
primary_slot_name = 'repmgr_slot_1'
```
It seems that replication is working fine in the sense of data correctness, but for some reason some  *_fsm files are not synced.
So here is the question:
- is it normal and it is pgbackrest issue that it confuses source of *_fsm files (i.e. it should take them only from primary or only form replica)
- or something is messed up with my replica? If so - what can it by, why it occurred and how can I fix that?

**Comment by dwsteele** [comment]
> so long that it undermines sense of using incremental instead of full

This is not entirely true -- it will still be faster and produce a small backup.

> is it normal and it is pgbackrest issue that it confuses source of *_fsm files

These files should be replicated and are always copied from the replica.

> or something is messed up with my replica?

Looks that way, but there might be something else at play. I would make a backup from the primary and resync the replica and then see if this happens again.

Do you have data_checksums enabled on the cluster?

**Comment by krmozejko** [comment]
> This is not entirely true -- it will still be faster and produce a small backup.

I should add "in our case". In general case you are right and it was overstatement.

> Looks that way, but there might be something else at play. I would make a backup from the primary and resync the replica and then see if this happens again.

I will try that.

> Do you have data_checksums enabled on the cluster?

No, they are off.

**Comment by dwsteele** [comment]
We have seen a few reports of this and finally realized it was a bug, fixed in PR #2124. Note that there is no backup integrity issue, but of course the longer backup is not ideal.

@krmozejko How would you like your name to appear in the release notes?
